internal/statemachine/helper.go

```diff
@@ -1002,42 +1015,121 @@ func (stateMachine *StateMachine) updateGrub(rootfsVolName string, rootfsPartNum
 			relpath:  "/sys",
 			typ:      "sysfs",
 		},
+		{
+			basePath: mountDir,
+			relpath:  "/run",
+			bind:     true,
+		},
 	}
 
-	for _, mp := range mountPoints {
-		mountCmds, umountCmds, err := mp.getMountCmd()
-		if err != nil {
-			return fmt.Errorf("Error preparing mountpoint \"%s\": \"%s\"",
-				mp.relpath,
-				err.Error(),
-			)
-		}
-		updateGrubCmds = append(updateGrubCmds, mountCmds...)
-		teardownCmds = append(umountCmds, teardownCmds...)
+	mountCmds, umountCmds, err := generateMountPointCmds(mountPoints, stateMachine.tempDirs.scratch)
+	if err != nil {
+		return err
 	}
 
+	setupGrubCmds = append(setupGrubCmds, mountCmds...)
+	teardownCmds = append(umountCmds, teardownCmds...)
+
 	teardownCmds = append([]*exec.Cmd{
 		execCommand("udevadm", "settle"),
 	}, teardownCmds...)
 
+	// udev needed to have grub-install properly work
+	grubPackages = append(grubPackages, "udev")
+
+	setupGrubCmds = append(setupGrubCmds,
+		aptInstallCmd(mountDir, grubPackages, false),
+		execCommand("chroot",
+			mountDir,
+			"grub-install",
+			loopUsed,
+			"--boot-directory=/boot",
+			"--efi-directory=/boot/efi",
+			fmt.Sprintf("--target=%s", target),
+			"--uefi-secure-boot",
+			"--no-nvram",
+		),
+	)
+
+	if architecture == arch.AMD64 {
+		setupGrubCmds = append(setupGrubCmds,
+			execCommand("chroot",
+				mountDir,
+				"grub-install",
+				loopUsed,
+				"--target=i386-pc",
+			),
+		)
+	}
+
 	divert, undivert := divertOSProber(mountDir)
 
-	updateGrubCmds = append(updateGrubCmds, divert)
+	setupGrubCmds = append(setupGrubCmds, divert)
 	teardownCmds = append([]*exec.Cmd{undivert}, teardownCmds...)
 
-	// actually run update-grub
-	updateGrubCmds = append(updateGrubCmds,
+	setupGrubCmds = append(setupGrubCmds,
 		execCommand("chroot",
 			mountDir,
 			"update-grub",
 		),
 	)
 
 	// now run all the commands
-	err = helper.RunCmds(updateGrubCmds, stateMachine.commonFlags.Debug)
+	return helper.RunCmds(setupGrubCmds, stateMachine.commonFlags.Debug)
+}
+
+func (stateMachine *StateMachine) confFromArch(architecture string) (string, []string) {
+	switch architecture {
+	case arch.AMD64:
+		return "x86_64-efi", []string{"grub-pc", "shim-signed"}
```

Comment:
This seems slightly incoherent, and in a way that might possibly even be interesting. grub-pc is only about BIOS boot. If you build an amd64 image and your gadget says the bootloader is grub, do we build an image that can only boot EFI or one that can boot grub as well? Ubuntu Core doesn't have this issue because it doesn't boot BIOS ever but for classic I don't suppose we can get away with that.

Sadly, we probably want to do both BIOS+UEFI and UEFI-only images (he says, desperately hoping there is not a use case for BIOS-only images out there).

For UEFI+BIOS images we'll need to call grub-install twice.

For now, it probably makes sense to build UEFI-only images. In which case you can remove grub-pc here.

Reply:
I think we need to support the UEFI+BIOS hybrid images from the start, sadly. IIRC the default pc classic gadget we always provided as a reference basically was a hybrid one. So not caring from the start I think would regress the behavior of the tool.
For BIOS-only images I had a chat with Paul earlier and we said we won't explicitly support just yet, but wait if someone asks for them. Since in theory the gadget.yaml allows that.

I'm thinking about what you said about calling grub-install twice, and I think I need to process that for a bit.

Reply:
I understand adding grub-pc and shim-signed will lead to a UEFI+BIOS image, right?

I am thinking we could also decide the "kind" of bootloader conf we build based on the EFI partition number found:
- If this is `0` (no EFI partition found) then we can/should build a BIOS-only image. In this case we only install `grub-pc` and we call grub-install without the efi-specifc args (like in a hook of livecd-rootfs)
- If this is not `0` (a EFI partition was found) then we build a UEFI+BIOS image. In this case we install `grub-pc` and `shim-signed`.

It remains the UEFI-only case for which I have no solution at the moment. Maybe something in the `gadget.yaml` could let users express that.

Reply:
UEFI+BIOS from the start > ok. maybe we support only that then?

> I understand adding grub-pc and shim-signed will lead to a UEFI+BIOS image, right?

Well to be pedantic, I'm not sure that having both of those packages installed means that the image will be bootable via both UEFI and BIOS without further steps, the bootloader assets also need to be installed into the ESP / boot sector of the drive, and relatedly-but-not-the same I don't know that having these packages installed means that those packages being upgraded means that the boot assets get updated in the ESP / boot sector -- which rather gets to the heart of the PR. Maybe we can confer with @kukrimate to ensure we have a solid understanding of what is going on -- I certainly find grub's postinst pretty confusing.

Reply:
And yes, I guess the UEFI vs BIOS vs UEFI+BIOS should be communicated via the gadget somehow. But also maybe we only support UEFI+BIOS for now.